From b5268745e427cbe650ba9da2e6bea2d80d3d7021 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lasse=20Brudeskar=20Vik=C3=A5s?= <bvx89.at.cs@gmail.com>
Date: Sun, 19 Jan 2014 22:01:57 +0100
Subject: [PATCH] Add Download Speeds in notification

This commit adds the ability to show the current total download
speed in the expanded mode. This is applicable to both single
and multi file downloads.

Screenshots:
Multi  - http://i.imgur.com/3t7yAy7.png
Single - http://i.imgur.com/MMqmA22.png

from slim rom

Change-Id: I801dbe61c7ee59d0c1d14d5851ad6dc3a7678499
---
 res/values/cm_strings.xml                          |  3 ++
 .../providers/downloads/DownloadNotifier.java      | 35 ++++++++++++++++++----
 2 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index e9409a9..d3e138e 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -17,4 +17,7 @@
 <resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
     <string name="live_folder_drawer_label">Downloads folder</string>
     <string name="live_folder_label">Downloads</string>
+
+    <!-- Transfer speed for current download/s. -->
+    <string name="download_speed"><xliff:g id="duration">%s</xliff:g>/s</string>
 </resources>
diff --git a/src/com/android/providers/downloads/DownloadNotifier.java b/src/com/android/providers/downloads/DownloadNotifier.java
index ac52eba..6c3c90a 100644
--- a/src/com/android/providers/downloads/DownloadNotifier.java
+++ b/src/com/android/providers/downloads/DownloadNotifier.java
@@ -35,6 +35,7 @@ import android.os.SystemClock;
 import android.provider.Downloads;
 import android.text.TextUtils;
 import android.text.format.DateUtils;
+import android.text.format.Formatter;
 import android.util.Log;
 import android.util.LongSparseLongArray;
 
@@ -202,6 +203,7 @@ public class DownloadNotifier {
             // Calculate and show progress
             String remainingText = null;
             String percentText = null;
+            String speedText = null;
             if (type == TYPE_ACTIVE) {
                 long current = 0;
                 long total = 0;
@@ -221,6 +223,10 @@ public class DownloadNotifier {
                     percentText = res.getString(R.string.download_percent, percent);
 
                     if (speed > 0) {
+                        // use Formatter interface for determining speed unit
+                        speedText = res.getString(R.string.download_speed,
+                                Formatter.formatFileSize(mContext, speed));
+
                         final long remainingMillis = ((total - current) * 1000) / speed;
                         remainingText = res.getString(R.string.download_remaining,
                                 DateUtils.formatDuration(remainingMillis));
@@ -235,16 +241,25 @@ public class DownloadNotifier {
             // Build titles and description
             final Notification notif;
             if (cluster.size() == 1) {
+                final Notification.InboxStyle inboxStyle = new Notification.InboxStyle(builder);
                 final DownloadInfo info = cluster.iterator().next();
 
                 builder.setContentTitle(getDownloadTitle(res, info));
+                builder.setContentText(remainingText);
 
                 if (type == TYPE_ACTIVE) {
                     if (!TextUtils.isEmpty(info.mDescription)) {
-                        builder.setContentText(info.mDescription);
+                        inboxStyle.addLine(info.mDescription);
+                    } else {
+                        inboxStyle.addLine(res.getString(R.string.download_running));
+                    }
+
+                    if (TextUtils.isEmpty(speedText)) {
+                        inboxStyle.setSummaryText(remainingText);
                     } else {
-                        builder.setContentText(remainingText);
+                        inboxStyle.setSummaryText(speedText + ", " + remainingText);
                     }
+
                     builder.setContentInfo(percentText);
 
                 } else if (type == TYPE_WAITING) {
@@ -253,14 +268,19 @@ public class DownloadNotifier {
 
                 } else if (type == TYPE_COMPLETE) {
                     if (Downloads.Impl.isStatusError(info.mStatus)) {
-                        builder.setContentText(res.getText(R.string.notification_download_failed));
+                        builder.setContentText(
+                                res.getText(R.string.notification_download_failed));
+                        inboxStyle.setSummaryText(
+                                res.getText(R.string.notification_download_failed));
                     } else if (Downloads.Impl.isStatusSuccess(info.mStatus)) {
                         builder.setContentText(
                                 res.getText(R.string.notification_download_complete));
+                        inboxStyle.setSummaryText(
+                                res.getText(R.string.notification_download_complete));
                     }
                 }
 
-                notif = builder.build();
+                notif = inboxStyle.build();
 
             } else {
                 final Notification.InboxStyle inboxStyle = new Notification.InboxStyle(builder);
@@ -274,7 +294,12 @@ public class DownloadNotifier {
                             R.plurals.notif_summary_active, cluster.size(), cluster.size()));
                     builder.setContentText(remainingText);
                     builder.setContentInfo(percentText);
-                    inboxStyle.setSummaryText(remainingText);
+
+                    if (TextUtils.isEmpty(speedText)) {
+                        inboxStyle.setSummaryText(remainingText);
+                    } else {
+                        inboxStyle.setSummaryText(speedText + ", " + remainingText);
+                    }
 
                 } else if (type == TYPE_WAITING) {
                     builder.setContentTitle(res.getQuantityString(
-- 
1.9.0

